### ISSUE TITLE
Possible disk space exhausted when running Docker on BTRFS by read-only traversing into /var/lib/docker/btrfs

### ISSUE THREAD

**Issue by mgtsai** [moby/moby#39815]
**Description**

When running Docker on BTRFS, Docker manages images by using many advantage features of BTRFS.  One of the important features is CoW (copy-on-write) to minimize disk usage.

But when the disk free space less than 15%, there is a certain probability that the following scenario may happen:

"Read-only traversing" into /var/lib/docker/btrfs/subvolumes causes "No space left on device".  Even the Docker daemon is stopped (ensure that no one is writing /var/lib/docker), the disk space may be drained by traversing into /var/lib/docker/btrfs/subvolumes.

"Read-only traversing" means doing read-only operations for the whole /var/lib/docker/btrfs/subvolumes tree, such as "du", "ls", "find", "rsync", "tar", ...

Ideally, read-only operations modify nothing in /var/lib/docker/btrfs/subvolumes, but it is not the case in reality.  While reading a file/directory, the atime (access time) of this file/directory would be modified, that is, the content of the parent directory containing the stat info of this file/directory would also be modified.

Docker manages images by using the CoW feature of BTRFS to minimize the disk usage, and most of the contents (including files and directories) in /var/lib/docker/btrfs/subvolumes are shared across different BTRFS subvolumes with high level duplication.  But when doing read-only traversing operations into /var/lib/docker/btrfs/subvolumes, it will trigger CoW and consumes disk space to record new atime of the traversed files/directories.  The amount of the disk space is largely consumed due to high level duplication

After doing some experiments, when disk usage is above 85%, the probability of draining disk space only by read-only traversing operations becomes higher. The remaining 15% free space may be drained by read-only traversing operations, such as "du", "ls", "find", "rsync", "tar", ... etc.  It is not expected by most system administrators especially when the disk space seems being still very sufficient at that moment.  This unexpected behavior will trouble many system administrators while doing normal regular maintenance.

**Steps to reproduce the issue:**

To highlight the scenario, a relative large image from Dockerhub is used to consume more space during traversing /var/lib/docker/btrfs/subvolumes.

1. Stop Docker daemon if it is running

2. Create a BTRFS image file of 5GB, and mount to /var/lib/docker
```
dd if=/dev/zero of=lookback.img bs=1048576 count=5120
mkfs.btrfs lookback.img
mount -o loop lookback.img /var/lib/docker
```

3. Start Docker daemon

4. Build an Docker image with Dockerfile of the following content:
```
FROM jupyter/scipy-notebook:latest
RUN echo 001 >> test
RUN echo 002 >> test
RUN echo 003 >> test
RUN echo 004 >> test
RUN echo 005 >> test
RUN echo 006 >> test
RUN echo 007 >> test
RUN echo 008 >> test
RUN echo 009 >> test
RUN echo 010 >> test
RUN echo 011 >> test
RUN echo 012 >> test
RUN echo 013 >> test
RUN echo 014 >> test
RUN echo 015 >> test
RUN echo 016 >> test
RUN echo 017 >> test
RUN echo 018 >> test
RUN echo 019 >> test
RUN echo 020 >> test
```
* Image jupyter/scipy-notebook:latest is large enough for demonstration
* Building a new image with 20 steps causes 20 more subvolumes in /var/lib/docker/btrfs/subvolumes, and will consume 20x more disk space

5. Stop Docker daemon -- Ensure no one is modifying /var/lib/docker

6. touch /var/lib/docker/before
- It successfully creates a new empty file "before"

7. du /var/lib/docker/btrfs/subvolumes
- Traverse into /var/lib/docker/btrfs/subvolumes

8. touch /var/lib/docker/after
- It fails by displaying: "touch: cannot touch '/var/lib/docker/after': No space left on device"

**Describe the results you received:**

After "du /var/lib/docker/btrfs/subvolumes" with doing nothing else, then "touch /var/lib/docker/after" causes:

"touch: cannot touch '/var/lib/docker/after': No space left on device"

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        74b1e89e8a
 Built:             Thu Jul 25 21:21:35 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89e8a
  Built:            Thu Jul 25 21:20:09 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 21
 Server Version: 19.03.1
 Storage Driver: btrfs
  Build Version: Btrfs v4.4
  Library Version: 101
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.4.0-1049-aws
 Operating System: Ubuntu 16.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 1
 Total Memory: 1.952GiB
 Name: aws-docker-testing
 ID: IHAF:W3NS:4TVP:3MHP:DAEJ:DMJM:DLW3:E4UD:M7AK:EEQA:PJWT:R5O4
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
  provider=amazonec2
 Experimental: false
 Insecure Registries:
  jeoutai.registry:5000
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by mgtsai** [comment]
With additional related long-lived issue #27653, I suggest the disk usage model should be further investigated and analyzed of image/container content structure while running Docker on BTRFS, especially for the influences of CoW.

**Comment by ViceIce** [comment]
I've had the same issue. Now trying to use `noatime` mount option, which hopefully resolves this error

**Comment by hebecked** [comment]
I have the same issue. 
Could I just delete files in "/var/lib/docker/btrfs/subvolumes" on a regular basis ?

**Comment by zilexa** [comment]
I now have the same issue. I guess lots of people and businesses use BTRFS and Docker. What am I missing? How to prevent this from happening?

**Comment by mmoya** [comment]
> I've had the same issue. Now trying to use `noatime` mount option, which hopefully resolves this error

@viceice Are using a dedicated volume for mounting `/var/lib/docker`? Can you elaborate more on how/where to use the `noatime` option?

**Comment by viceice** [comment]
Yes, my root is ext4 ubuntu and I've a lvm btrfs mount to `/var/lib/docker/`. Since `noatime` I've never had that issue.

But now I switched to kubernetes with xfs.
